[Extensions] Detect gift card extension

ThomazFB opened the pull request:
Summary
==========
Adds validation to the Order Creation/Edit view that ensures that the `Add gift card` button is only available when the Gift card extension is installed in the store.

How to Test
==========
### Scenario 1 - Store with Gift Card installed

Prerequisite: the store has the [Gift Cards](https://woocommerce.com/products/gift-cards/) extension/plugin.

- Launch the app
- Go to the Orders tab
- Tap `+` FAB to create an order
- Add a product, or set a fee so that the order total is non-zero
- Verify that the `Add Gift Card` is visible and interactable


### Scenario 1 - Store with Gift Card installed

Prerequisite: the store doesn't have the Gift cards extension installed.

- Launch the app
- Go to the Orders tab
- Tap `+` FAB to create an order
- Add a product, or set a fee so that the order total is non-zero
- Verify that the `Add Gift Card` is not visible

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Introduce OrderCreateEditRepository.isGiftCardExtensionEnabled function

ThomazFB committed:
Add unit test coverage for OrderCreateEditRepository.isGiftCardExtensionEnabled function

ThomazFB committed:
Refactor OrderCreateEditRepository gift card plugin info to fetch a full plugin list

ThomazFB committed:
Adjust OrderCreateEditRepository to return a map of plugins associated with it name instead of al ist

ThomazFB committed:
Update OrderCreateEditRepositoryTest

ThomazFB committed:
Refactor plugin fetch to correctly fetch a map of WooPlugins

ThomazFB committed:
Add a pluginsInformation field to contain all supported plugin data inside the Order Create Edit view

ThomazFB committed:
Add gift card extension detection to OrderCreateEditViewModel

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Configure add gift card button visibility to account for extension availability

ThomazFB committed:
Merge branch 'issue/validate-gift-card-format' into issue/detect-gift-card-extension

atorresveiga commented:
@ThomazFB WDYT if we hide the gift card by default and only show the Add gift card button when the extension is installed? 

I'm not convinced about the experience for sites without the extension of showing the button first disabled and disappearing after a product is added. 



https://github.com/woocommerce/woocommerce-android/assets/18119390/fa540195-cb9d-456a-8f37-4caab1f0cf5d

atorresveiga commented:
I found one issue @ThomazFB unrelated to the changes in this PR, but that happens with this code.
If I try to create an order without a gift card code, the operation fails.

https://github.com/woocommerce/woocommerce-android/assets/18119390/89c0ea78-7799-4d86-87f8-96cb8f2f2690

This is the error on Flipper:

![image](https://github.com/woocommerce/woocommerce-android/assets/18119390/f618cdc5-bf95-4b86-8e16-f74abc4410de)

ThomazFB committed:
Merge branch 'trunk' into issue/detect-gift-card-extension

atorresveiga commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditRepositoryTest.kt:
```diff
@@ -209,4 +211,29 @@ class OrderCreateEditRepositoryTest : BaseUnitTest() {
+    fun `when isGiftCardExtensionEnabled is called, then it should return the correct value`() = testBlocking {
+        // Given
+        val pluginMock = mock<SitePluginModel> {
+            on { name } doReturn "test plugin"
```
np: WDYT if we use the real plugin name here?

ThomazFB replied:
Sure! Fixed in https://github.com/woocommerce/woocommerce-android/pull/10377/commits/efb1b9593c601c29ee56a5dbd2b6b6cbeb3ca77e

atorresveiga commented:
@ThomazFB, is it possible that this PR depends on a FluxC PR?

I'm getting this error now when trying to test the PR:

> java.lang.IllegalStateException: Room cannot verify the data integrity. Looks like you've changed schema but forgot to update the version number. You can simply fix this by increasing the version number. Expected identity hash: 892bffac2c3dc56265f35a4375208c54, found: a8d05c3a70fd491acc9dd8dbc2c8bce3

ThomazFB commented:
Hey @atorresveiga, I think I know the reason for this error: in the last FluxC change, I updated the schema 32 I introduced there. At first, it contained only the gift card code field, but now it also includes the gift card discount amount. If your device runs the previous version of this code with the previous schema 32, it will probably conflict because it's the same schema number but a different schema nonetheless.

The change I'm talking about is in this commit: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2905/commits/6be7dc898034a6bc7e248b63a5f69c146dc5ab74. I updated the schema 32, so it's possible that your current installed version is running the schema 32 from that PR before my update, but the JSON file describes a different one, but since they're the same number, Room thinks it's a new version that forgot to have the number increased.

To make sure we're still with database integrity, I would like to ask you to do the following steps:
1 - Uninstall the Woo app
2 - Install the app from the trunk branch
3 - Install the app from this branch on top of the trunk installation to run the migration
4 - Check if the error persists

ThomazFB committed:
Merge branch 'trunk' into issue/detect-gift-card-extension

atorresveiga commented:
> I updated the schema 32, so it's possible that your current installed version is running the schema 32 from that PR before my update, but the JSON file describes a different one, but since they're the same number, Room thinks it's a new version that forgot to have the number increased.

It was this. ☝️ 

> To make sure we're still with database integrity, I would like to ask you to do the following steps:
1 - Uninstall the Woo app
2 - Install the app from the trunk branch
3 - Install the app from this branch on top of the trunk installation to run the migration
4 - Check if the error persists

This fixed the issue I was facing, thanks for the quick response

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditFormFragment.kt:
```diff
@@ -740,7 +740,7 @@ class OrderCreateEditFormFragment :
         orderEditGiftCardLayout.hide()
         if (newOrderData.selectedGiftCard.isNullOrEmpty()) {
-            addGiftCardButton.isVisible = true
+            addGiftCardButton.isVisible = viewModel.isGiftCardExtensionEnabled
```
Because of the condition in line `605` when we try to create an order the first time, the code does not reach the bindPaymentSection -> bindGiftCardSubSection, and the add gift card button remains visible until we add either a fee or an item.
```
 if (newOrderData.items.isEmpty() && newOrderData.feesLines.isEmpty()) 
```
Maybe the solution could be to observe `isGiftCardExtensionEnabled` and display/hide the button based on that. WDYT?

ThomazFB replied:
I replied this suggestion in the main PR thread to keep all information in one place: https://github.com/woocommerce/woocommerce-android/pull/10377#issuecomment-1865291441

ThomazFB committed:
Update FluxC version to trunk checkout

ThomazFB committed:
Add GiftCard extension availability control to OrderCreateEditViewModel

ThomazFB committed:
Extract plugin data monitoring to dedicated function

ThomazFB committed:
Add reactive control of gift card button visibility

ThomazFB committed:
Fix lint issues

ThomazFB commented:
Hey @atorresveiga! Thanks for your suggestions! I 100% agree with your `add gift button` visibility point, and I have an update on this: I had to do quite a rework to meet the proper `add gift card` button visibility bound to the plugin availability. Even thought you pointed out a problem with the UI refresh [here](https://github.com/woocommerce/woocommerce-android/pull/10377#discussion_r1432188880), we had an race condition issue, where fetching the plugin availability information from the database and the UI refresh where running at the same time.

To properly solve this, I implemented a reactive approach with a `ViewState` update. We already have the `isAddGiftCardButtonEnabled` controlling if the add gift card button is enabled or not, so we avoid enabling it during updates. But since the Plugin availability controls the `visibility` of the add gift card button, I decided to introduce a `shouldDisplayAddGiftCardButton` field that is controlled by a StateFlow of the Plugin information, this way, we kill the race condition and ensure a consistent UI control.

This PR is ready for another round, let me know what you think of this fix!

ThomazFB committed:
Avoid internal reference in state flow reactions

ThomazFB committed:
Update OrderCreateEditRepositoryTest

atorresveiga commented:
Thanks for working on the button visibility 🙏 
Great work, Thomaz! 🎉 
Everything is working as expected! 

LGTM! :shipit:
